File: progminerlab/disasm.py
#!/usr/bin/env python3

import sys


def reg(x):
    return reg.REGISTERS[x]


reg.REGISTERS = {
    0x0: 'R0',
    0x1: 'R1',
    0x2: 'R2',
    0x3: 'R3',
    0x4: 'R4',
    0x5: 'R5',
    0x6: 'R6',
    0x7: 'R7',
    0x8: 'R8',
    0x9: 'R9',
    0xA: 'R10',
    0xB: 'R11',
    0xC: 'R12',
    0xD: 'TMP',
    0xE: 'SP',
    0xF: 'FP',
}


def disasm(code, labels=dict(), pos=0):
    '''
    code - file-like object of bytes to disassemble
    labels - dict from offset in code to label
    '''

    def read_small_const(f=lambda x: f'{x}'):
        nonlocal instr

        x = code.read(1)

        instr += x
        x = int.from_bytes(x, signed=True)

        return f(x)

    def read_large_const(f=lambda x: f'{x}'):
        nonlocal instr

        x = code.read(4)

        instr += x
        x = int.from_bytes(x, 'little', signed=True)

        return f(x)

    def read_offset():
        return read_small_const(lambda x: f'{x} -> 0x{pos + x:04X}')

    def read_regs(f=lambda r1, r2: f'{r1}, {r2}'):
        nonlocal instr

        regs = code.read(1)

        instr += regs

        regs = regs[0]
        reg1, reg2 = regs >> 4, regs & 0xF

        return f(reg(reg1), reg(reg2))

    while opcode := code.read(1):
        opcode = opcode[0]

        instr = [opcode]
        mnemo = None

        match [opcode >> 4, opcode & 0xF]:
            case [0x0, 0x0]:
                mnemo = 'NOP'

            case [0x0, 0x1]:
                mnemo = f'JMPs {read_offset()}'

            case [0x0, 0x2]:
                mnemo = 'FLUSH'

            case [0x0, 0x3]:
                mnemo = f'CALL {read_offset()}'

            case [0x0, 0x4]:
                mnemo = 'RET'

            case [0x1, 0x0]:
                mnemo = f'MOV {read_regs()}'

            case [0x1, 0x1]:
                mnemo = f'ADD {read_regs()}'

            case [0x1, 0x2]:
                mnemo = f'SUB {read_regs()}'

            case [0x1, 0x3]:
                mnemo = f'MUL {read_regs()}'

            case [0x1, 0x4]:
                mnemo = f'DIV {read_regs()}'

            case [0x1, 0x5]:
                mnemo = f'REM {read_regs()}'

            case [0x1, 0x6]:
                mnemo = f'AND {read_regs()}'

            case [0x1, 0x7]:
                mnemo = f'OR {read_regs()}'

            case [0x1, 0x8]:
                mnemo = f'XOR {read_regs()}'

            case [0x1, 0x9]:
                mnemo = f'SHL {read_regs()}'

            case [0x1, 0xA]:
                mnemo = f'SHR {read_regs()}'

            case [0x1, 0xB]:
                mnemo = f'ASR {read_regs()}'

            case [0x1, 0xC]:
                mnemo = f'LOAD {read_regs()}'

            case [0x1, 0xD]:
                mnemo = f'STORE {read_regs()}'

            case [0x1, 0xE]:
                mnemo = f'PIXEL_OFF {read_regs()}'

            case [0x1, 0xF]:
                mnemo = f'PIXEL_ON {read_regs()}'

            case [0x2, 0x1]:
                mnemo = f'CMPeq {read_regs()}'

            case [0x2, 0x2]:
                mnemo = f'CMPgt {read_regs()}'

            case [0x2, 0x3]:
                mnemo = f'CMPge {read_regs()}'

            case [0x2, 0x4]:
                mnemo = f'CMPlt {read_regs()}'

            case [0x2, 0x5]:
                mnemo = f'CMPle {read_regs()}'

            case [0x2, 0x6]:
                mnemo = f'CMPne {read_regs()}'

            case [0x3, r]:
                mnemo = f'CONSTs {reg(r)}, {read_small_const()}'

            case [0x4, r]:
                mnemo = f'CONSTl {reg(r)}, {read_large_const()}'

            case [0x5, r]:
                mnemo = f'IFZs {reg(r)}, {read_offset()}'

            case [0x6, r]:
                mnemo = f'RAND {reg(r)}'

            case _:
                mnemo = f'# UNKNOWN {hex(opcode)}'

        instrStr = ' '.join([f'{x:02X}' for x in instr])
        print(f'{pos:04X} | {instrStr:<14} | {mnemo}')

        pos += len(instr)


if __name__ == '__main__':
    if len(sys.argv) > 1:
        # TODO support ELF parsing with labels

        print(f'Usage: {sys.argv[0]}')
        exit(0)

    disasm(sys.stdin.buffer)

File: progminerlab/src/disasm.cpp
#include <sstream>
#include <iomanip>

#include "loader.h"


struct disasm_info : code_info {

    std::unordered_map<std::size_t, std::string> labels;

    void add_labels();

    void print_offset(std::ostream & os, std::size_t pos, const uint8_t * it) const;

    void disasm(std::ostream &) const;
};

void disasm_info::add_labels() {
    for (const uint8_t * it = code_begin; it < code_end; ) {
        const uint8_t * const next_it = next_instr(it);

        const uint8_t * tmp = it;
        const auto opcode = read_number<uint8_t>(tmp);

        if (opcode == 0x01 || opcode == 0x03 || (opcode >> 4) == 0x5) {
            const int32_t off = read_number<int32_t>(tmp);

            const std::size_t pos = std::distance(code_begin, it) + off;
            if (auto it_label = labels.find(pos); it_label == labels.end()) {
                std::ostringstream os;
                os << "L_" << pos;

                labels[pos] = os.str();
            }
        }

        it = next_it;
    }
}

static void print_byte(std::ostream & os, uint8_t b) {
    const char pf = os.fill('0');

    os << std::setw(2) << std::hex << std::uppercase << static_cast<uint32_t>(b)
       << std::dec << std::nouppercase;

    os.fill(pf);
}

static void print_pos(std::ostream & os, std::size_t pos) {
    const char pf = os.fill('0');

    os << std::setw(4) << std::hex << std::uppercase << pos
       << std::dec << std::nouppercase;

    os.fill(pf);
}

void disasm_info::print_offset(std::ostream & os, std::size_t pos, const uint8_t * it) const {
    const auto off = read_number<int32_t>(it);

    pos += off;
    os << off << " -> ";

    if (auto it_label = labels.find(pos); it_label != labels.end()) {
        os << it_label->second;
    } else {
        print_pos(os, pos);
    }
}

const char * reg_to_string(uint8_t reg) {
    static const char * const REGISTERS[] = {
        "R0", "R1", "R2", "R3", "R4", "R5", "R6", "R7", "R8", "R9",
        "R10", "R11", "R12", "TMP", "SP", "FP"
    };

    return REGISTERS[reg & 0xF];
}

void print_reg(std::ostream & os, uint8_t reg) {
    os << reg_to_string(reg);
}

void print_reg2(std::ostream & os, const uint8_t * it) {
    const uint8_t regs = *it;

    print_reg(os, regs >> 4);
    os << ", ";

    print_reg(os, regs & 0xF);
}

void disasm_info::disasm(std::ostream & os) const {
    for (const uint8_t * it = code_begin; it < code_end; ) {
        const uint8_t * const next_it = next_instr(it);

        const std::size_t pos = std::distance(code_begin, it);

        if (auto it_label = labels.find(pos); it_label != labels.end()) {
            os << std::endl;

            os << it_label->second << ':' << std::endl;
        }

        print_pos(os, pos);
        os << " | ";

        for (const uint8_t * it1 = it; it1 < next_it; ++it1) {
            print_byte(os, *it1);
            os << ' ';
        }

        for (std::size_t i = std::distance(it, next_it); i < 5; ++i) {
            os << "   ";
        }

        os << "| ";

        const uint8_t * tmp = it;
        const auto opcode = read_number<uint8_t>(tmp);

        switch (opcode >> 4) {
        case 0x0:
            switch (opcode & 0xF) {
            case 0x0:
                os << "NOP";
                break;

            case 0x1:
                os << "JMP ";
                print_offset(os, pos, tmp);
                break;

            case 0x2:
                os << "FLUSH";
                break;

            case 0x3:
                os << "CALL ";
                print_offset(os, pos, tmp);
                break;

            case 0x4:
                os << "RET";
                break;

            default:
                goto unknown;
            }

            break;

        case 0x1:
            switch (opcode & 0xF) {
            case 0x0:
                os << "MOV ";
                print_reg2(os, tmp);
                break;

            case 0x1:
                os << "ADD ";
                print_reg2(os, tmp);
                break;

            case 0x2:
                os << "SUB ";
                print_reg2(os, tmp);
                break;

            case 0x3:
                os << "MUL ";
                print_reg2(os, tmp);
                break;

            case 0x4:
                os << "DIV ";
                print_reg2(os, tmp);
                break;

            case 0x5:
                os << "REM ";
                print_reg2(os, tmp);
                break;

            case 0x6:
                os << "AND ";
                print_reg2(os, tmp);
                break;

            case 0x7:
                os << "OR ";
                print_reg2(os, tmp);
                break;

            case 0x8:
                os << "XOR ";
                print_reg2(os, tmp);
                break;

            case 0x9:
                os << "SHL ";
                print_reg2(os, tmp);
                break;

            case 0xA:
                os << "SHR ";
                print_reg2(os, tmp);
                break;

            case 0xB:
                os << "ASR ";
                print_reg2(os, tmp);
                break;

            case 0xC:
                os << "LOAD ";
                print_reg2(os, tmp);
                break;

            case 0xD:
                os << "STORE ";
                print_reg2(os, tmp);
                break;

            case 0xE:
                os << "PIXEL_OFF ";
                print_reg2(os, tmp);
                break;

            case 0xF:
                os << "PIXEL_ON ";
                print_reg2(os, tmp);
                break;
            }

            break;

        case 0x2:
            switch (opcode & 0xF) {
            case 0x1:
                os << "CMPeq ";
                print_reg2(os, tmp);
                break;

            case 0x2:
                os << "CMPgt ";
                print_reg2(os, tmp);
                break;

            case 0x3:
                os << "CMPge ";
                print_reg2(os, tmp);
                break;

            case 0x4:
                os << "CMPlt ";
                print_reg2(os, tmp);
                break;

            case 0x5:
                os << "CMPle ";
                print_reg2(os, tmp);
                break;

            case 0x6:
                os << "CMPne ";
                print_reg2(os, tmp);
                break;

            default:
                goto unknown;
            }

            break;

        case 0x3:
            os << "CONSTs ";
            print_reg(os, opcode);
            os << ", " << static_cast<int32_t>(read_number<int8_t>(tmp));
            break;

        case 0x4:
            os << "CONSTl ";
            print_reg(os, opcode);
            os << ", " << read_number<int32_t>(tmp);
            break;

        case 0x5:
            os << "IFZ ";
            print_reg(os, opcode);
            os << ", ";
            print_offset(os, pos, tmp);
            break;

        case 0x6:
            os << "RAND ";
            print_reg(os, opcode);
            break;

unknown:
        default:
            os << "# UNKNOWN";
            break;
        }

        os << std::endl;

        it = next_it;
    }
}

static disasm_info load_file(const ELFIO::elfio & reader) {
    check_file(reader);

    disasm_info result;
    find_code_section(reader, result);

    for (auto [name, pos] : read_symtab(reader)) {
        result.labels[pos] = name;
    }

    return result;
}

int main(int argc, char * argv[]) {
    if (argc != 2) {
        std::cout << "Usage: " << argv[0] << " <ELF file>" << std::endl;
        return 0;
    }

    ELFIO::elfio reader;
    if (!reader.load(argv[1])) {
        std::cerr << "Unable to read file " << argv[1] << " as ELF" << std::endl;
        return 1;
    }

    disasm_info info;

    try {
        info = load_file(reader);
    } catch (const std::exception & e) {
        std::cerr << "Bad ELF file: " << e.what() << std::endl;
        return 2;
    }

    info.add_labels();
    info.disasm(std::cout);
}
